polynomial.c

```diff
@@ -1,128 +1,157 @@
+  }
+  //DEBUG printf("local sum for rank %d is %f \n", rank,localSum);
+  
+  return localSum;
 }
 
-// Driver Code
-int main(int argc, char **argv)
-{
-    
-    int *coeffArr = (int *)malloc(sizeof(int) * MAX);
+double runSequential(int rank, int numPolynomials) {
+  if(rank == 0) {
+    //Each process generates the full set of coefficients.
+    //DEBUG printf("Initializing polynomials\n");
+    int *coeffArr = (int *)malloc(sizeof(int) * numPolynomials);
+    initialize(coeffArr, numPolynomials);
     
-    initialize(coeffArr);
-    double x = 0.99;
-    
-    MPI_Init (&argc, &argv);
-    int rank;
-    int numProcs;
-    MPI_Comm_rank(MPI_COMM_WORLD, &rank);  // rank is process id
-    MPI_Comm_size(MPI_COMM_WORLD, &numProcs); // number of processes in total, from command line 
-    
-
     /* Start timer */
     double elapsed_time = - MPI_Wtime();
-    
-    //double seq_value = sequential(coeffArr, x);
-    double chunk_value = chunk(coeffArr, x, rank, numProcs);
+
+    double localTotal = sequential(coeffArr, numPolynomials, VARIABLE);
+
+    if (rank == 0) {
+      printf("\n\nGlobal Total: %f\n", localTotal);
+    }
+
     /* End timer */
     elapsed_time = elapsed_time + MPI_Wtime();
-   
-   if(rank == 0){
-      //printf(" sequential value %f wall clock time %8.6f \n", seq_value, elapsed_time);
-      printf(" chunk value %f wall clock time %8.6f \n", chunk_value, elapsed_time);
-   }
-    
-    fflush(stdout);
-    
+    printf(" sequential value %f wall clock time %8.6f \n", localTotal, elapsed_time);  
+
     free(coeffArr);
+
+  }
+}
+
+double runDistributed(int rank, int numProcs, int numPolynomials, double(*multiThreadStrategy)(int*, int, int, int, double)) 
+{
+  //Each process generates the full set of coefficients.
+  //DEBUG printf("Initializing polynomials\n");
+  int *coeffArr = (int *)malloc(sizeof(int) * numPolynomials);
+  initialize(coeffArr, numPolynomials);
+  
+  /* Start timer */
+  double elapsed_time = - MPI_Wtime();
     
-    MPI_Finalize();
-    
-	return 0;
+  //Processes calculate work as local totals using round robin.
+  double localTotal = multiThreadStrategy(coeffArr, numPolynomials, rank, numProcs, VARIABLE);
+
+  //Aggregate results with reduce
+  double localResult = 0.0;
+  MPI_Reduce(&localTotal, &localResult, 1, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD); 
+
+  //result = aggregateMessagePassing(localTotal, rank, numProcs);
+
+  /* End timer */
+  elapsed_time = elapsed_time + MPI_Wtime();
+  
+  double maxElapsedTime;
+  MPI_Reduce(&elapsed_time, &maxElapsedTime, 1, MPI_DOUBLE, MPI_MAX, 0, MPI_COMM_WORLD);
+  if (rank == 0) 
+    printf(" sequential value %f wall clock time %8.6f \n", localResult, maxElapsedTime);  
+
+  free(coeffArr);
+
+}
+
+int main(int argc, char** argv) {
+  int rank;
+  int numProcs;
+  
+  MPI_Init(&argc, &argv);
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+  MPI_Comm_size(MPI_COMM_WORLD, &numProcs);
+
+  int numPolynomials = 50000;
+  if (argc >= 2) {
```

Comment:
I could probably be fancy here and do something where each param gets a tag like "-p" or whatever, but this does enough at least for me. Could probably also add a param for VARIABLE